core/src/subgraph/context/instance/mod.rs

```diff
@@ -127,28 +130,44 @@ where
         data_source: DataSource<C>,
     ) -> Result<Option<Arc<T::Host>>, Error> {
         // Protect against creating more than the allowed maximum number of data sources
-        if self.hosts.len() >= ENV_VARS.subgraph_max_data_sources {
+        if self.hosts_len() >= ENV_VARS.subgraph_max_data_sources {
             anyhow::bail!(
                 "Limit of {} data sources per subgraph exceeded",
                 ENV_VARS.subgraph_max_data_sources,
             );
         }
 
-        // `hosts` will remain ordered by the creation block.
-        // See also 8f1bca33-d3b7-4035-affc-fd6161a12448.
-        assert!(
-            self.hosts.last().and_then(|h| h.creation_block_number())
-                <= data_source.creation_block()
-        );
-
+        let is_onchain = data_source.is_onchain();
         let Some(host) = self.new_host(logger.clone(), data_source)? else { return Ok(None) };
 
-        Ok(if self.hosts.contains(&host) {
-            None
-        } else {
-            self.hosts.push(host.cheap_clone());
-            Some(host)
-        })
+        // Check for duplicates and add the host.
+        match is_onchain {
+            true => {
+                // `onchain_hosts` will remain ordered by the creation block.
+                // See also 8f1bca33-d3b7-4035-affc-fd6161a12448.
+                assert!(
+                    self.onchain_hosts
+                        .last()
+                        .and_then(|h| h.creation_block_number())
+                        <= host.data_source().creation_block()
+                );
```

Comment:
I am always a bit queasy with asserts since it's not that easy to spot panics in the logs; could we return an error here? That way, the subgraph will fail and the failure can tell use what exactly is wrong.